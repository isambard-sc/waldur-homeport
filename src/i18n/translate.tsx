import { Fragment, ReactNode } from 'react';

import { ENV } from '@waldur/configs/default';

import { LanguageUtilsService } from './LanguageUtilsService';
import { Translate } from './types';

export const formatJsxTemplate = (template, context) => {
  if (!context) {
    return template;
  }
  return (
    <Fragment>
      {template.split(/\{|\}/g).map((part, index) => (
        <Fragment key={index}>
          {index % 2 === 0 ? part : context[part]}
        </Fragment>
      ))}
    </Fragment>
  );
};

export const formatJsx = (
  template: string,
  context: Record<string, (s: string) => ReactNode>,
) => {
  const pattern = /<([^>]+)>([^<]*)<\/([^>]+)>/g;
  const parts = [];
  let matches,
    prevIndex = 0;
  while ((matches = pattern.exec(template)) !== null) {
    parts.push(template.substring(prevIndex, matches.index));
    parts.push(context[matches[1]](matches[2]));
    prevIndex = matches[0].length + matches.index;
  }
  if (prevIndex !== template.length) {
    parts.push(template.substring(prevIndex));
  }
  return (
    <Fragment>
      {parts.map((part, index) => (
        <Fragment key={index}>{part}</Fragment>
      ))}
    </Fragment>
  );
};

export const formatTemplate: Translate = (template, context) =>
  context ? template.replace(/{(.+?)}/g, (_, key) => context[key]) : template;

const translateTemplate = (template) =>
  LanguageUtilsService.dictionary[template] || template;

const getDomainMessage = (message) => {
  const domain = ENV.plugins?.WALDUR_CORE?.TRANSLATION_DOMAIN;
  if (!domain) {
    return message;
  }
  if (!DOMAIN_MESSAGES[domain]) {
    return message;
  }
  return DOMAIN_MESSAGES[domain][message] || message;
};

export const translate: Translate = (
  template,
  context,
  interpolator = formatTemplate,
) => interpolator(translateTemplate(getDomainMessage(template)), context);

const DOMAIN_MESSAGES = {
  service_catalogue: {
    'Explore {deployment} Marketplace': translate(
      'Explore {deployment} service catalog',
    ),
    'Go to marketplace': translate('Go to service catalog'),
    'Label that is visible to users in Marketplace.': translate(
      'Label that is visible to users in service catalog.',
    ),
    Marketplace: translate('Service catalog'),
    'Marketplace offerings': translate('Service catalog offerings'),
    'Welcome to marketplace': translate('Welcome to service catalog'),
    'Marketplace services can only be provisioned for a certain affiliation. You currently do not have any organizations or projects connected with your account.':
      translate(
        'Services can only be provisioned for a certain affiliation. You currently do not have any organizations or projects connected with your account.',
      ),
    'Register as a customer of our portal and provide your cloud services through the Marketplace.':
      translate(
        'Register as a customer of our portal and provide your cloud services through the service catalog.',
      ),
    'Resource is not connected to the marketplace yet.': translate(
      'Resource is not connected to the service catalog yet.',
    ),
    'There are no categories in marketplace yet.': translate(
      'There are no categories in catalog yet.',
    ),
    'There are no offerings in marketplace yet.': translate(
      'There are no offerings in catalog yet.',
    ),
    'Unique ID of a resource created via Marketplace': translate(
      'Unique ID of a resource created via service catalog',
    ),
    'You can find offerings to order in the marketplace': translate(
      'You can find offerings to order in the service catalog',
    ),
  },
  academic: {
    'Become a customer of our portal. Provision IT services from the Marketplace and manage your team from one place.':
      translate(
        'Become a customer of our portal. Provision IT services from the catalog and manage your team from one place.',
      ),
    'Explore {deployment} Marketplace': translate(
      'Explore {deployment} service catalog',
    ),
    'Go to marketplace': translate('Go to service catalog'),
    'Label that is visible to users in Marketplace.': translate(
      'Label that is visible to users in service catalog.',
    ),
    Marketplace: translate('Service catalog'),
    'Marketplace offerings': translate('Service catalog offerings'),
    'Welcome to marketplace': translate('Welcome to service catalog'),
    'Marketplace services can only be provisioned for a certain affiliation. You currently do not have any organizations or projects connected with your account.':
      translate(
        'Services can only be provisioned for a certain affiliation. You currently do not have any organizations or projects connected with your account.',
      ),
    'Register as a customer of our portal and provide your cloud services through the Marketplace.':
      translate(
        'Register as a customer of our portal and provide your cloud services through the service catalog.',
      ),
    'Resource is not connected to the marketplace yet.': translate(
      'Resource is not connected to the service catalog yet.',
    ),
    'There are no categories in marketplace yet.': translate(
      'There are no categories in catalog yet.',
    ),
    'There are no offerings in marketplace yet.': translate(
      'There are no offerings in catalog yet.',
    ),
    'Unique ID of a resource created via Marketplace': translate(
      'Unique ID of a resource created via service catalog',
    ),
    'You can find offerings to order in the marketplace': translate(
      'You can find offerings to order in the service catalog',
    ),
    Purchase: translate('Request'),
    'You have the right to purchase service without additional approval.':
      translate(
        'You have the right to request service without additional approval.',
      ),
  },
  academic_shared: {
    'Become a customer of our portal. Provision IT services from the Marketplace and manage your team from one place.':
      translate(
        'Become a customer of our portal. Provision IT services from the catalog and manage your team from one place.',
      ),
    'Explore {deployment} Marketplace': translate(
      'Explore {deployment} service catalog',
    ),
    'Go to marketplace': translate('Go to service catalog'),
    'Label that is visible to users in Marketplace.': translate(
      'Label that is visible to users in service catalog.',
    ),
    Marketplace: translate('Service catalog'),
    'Marketplace offerings': translate('Service catalog offerings'),
    'Welcome to marketplace': translate('Welcome to service catalog'),
    'Marketplace services can only be provisioned for a certain affiliation. You currently do not have any organizations or projects connected with your account.':
      translate(
        'Services can only be provisioned for a certain affiliation. You currently do not have any organizations or projects connected with your account.',
      ),
    'Register as a customer of our portal and provide your cloud services through the Marketplace.':
      translate(
        'Register as a customer of our portal and provide your cloud services through the service catalog.',
      ),
    'Resource is not connected to the marketplace yet.': translate(
      'Resource is not connected to the service catalog yet.',
    ),
    'There are no categories in marketplace yet.': translate(
      'There are no categories in catalog yet.',
    ),
    'There are no offerings in marketplace yet.': translate(
      'There are no offerings in catalog yet.',
    ),
    'Unique ID of a resource created via Marketplace': translate(
      'Unique ID of a resource created via service catalog',
    ),
    'You can find offerings to order in the marketplace': translate(
      'You can find offerings to order in the service catalog',
    ),
    Purchase: translate('Request'),
    'You have the right to purchase service without additional approval.':
      translate(
        'You have the right to request service without additional approval.',
      ),
  },
};
