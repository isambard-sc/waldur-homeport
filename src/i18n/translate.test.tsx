import { render } from '@testing-library/react';
import { describe, it, expect } from 'vitest';

import { formatJsx, formatJsxTemplate } from './translate';

describe('formatJsxTemplate', () => {
  it('allows to use curly braces syntax to interpolate JSX component', () => {
    const supportEmail = 'admin@example.com';
    const { container } = render(
      formatJsxTemplate('Please send an email to {supportEmail}. Thank you!', {
        supportEmail: <a href={`mailto:${supportEmail}`}>{supportEmail}</a>,
      }),
    );
    expect(container).toMatchSnapshot();
  });
});

describe('formatJsx', () => {
  it('allows to use angular braces syntax to interpolate JSX component', () => {
    const { container } = render(
      formatJsx(
        'By submitting the form you are agreeing to the <Link>Terms of Service</Link>.',
        {
          Link: (s) => <a href="example.com">{s}</a>,
        },
      ),
    );
    expect(container).toMatchSnapshot();
  });
});
