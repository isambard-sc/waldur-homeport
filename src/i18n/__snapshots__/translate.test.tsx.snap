// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`formatJsx > allows to use angular braces syntax to interpolate JSX component 1`] = `
<div>
  By submitting the form you are agreeing to the 
  <a
    href="example.com"
  >
    Terms of Service
  </a>
  .
</div>
`;

exports[`formatJsxTemplate > allows to use curly braces syntax to interpolate JSX component 1`] = `
<div>
  Please send an email to 
  <a
    href="mailto:admin@example.com"
  >
    admin@example.com
  </a>
  . Thank you!
</div>
`;
